### PR TITLE
libp2p_swarm_derive: Add `crate` parameter to configure `libp2p` crate name

### PR DESCRIPTION
**Pull request by stormshield-pj50**
# Description

When importing libp2p with a use statement and a custom alias name, the code generated by the NetworkBehaviour derive macro does not use this alias and does not compile. This PR enhances the derive macro with a new parameter "crate" that can be used to specify the custom alias name.

## Open Questions
None

## Change checklist

- [X] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm-derive/tests/test.rs [hunk 1]
```diff
@@ -133,6 +133,23 @@ fn custom_event() {
     }
 }
 
+#[test]
+fn custom_prefix() {
+    use libp2p as mylibp2p;
+    #[allow(dead_code)]
+    #[derive(NetworkBehaviour)]
+    #[behaviour(libp2p_prefix = "mylibp2p")]
```
Comment (thomaseizinger) [inline swarm-derive/tests/test.rs]: Serde calls this parameter `crate`: https://serde.rs/container-attrs.html#crate

I'd be in favor of following that here too.

Comment (stormshield-pj50) [reply swarm-derive/tests/test.rs]: OK, done

**Review by thomaseizinger — APPROVED** [review]
LGTM, thanks!

File: swarm-derive/src/lib.rs [hunk 2]
```diff
@@ -46,21 +46,43 @@ fn build(ast: &DeriveInput) -> TokenStream {
...
+            if let syn::NestedMeta::Meta(syn::Meta::NameValue(ref m)) = meta_item {
+                if m.path.is_ident("crate") {
+                    if let syn::Lit::Str(ref s) = m.lit {
+                        return Some(syn::parse_str::<syn::TypePath>(&s.value()).unwrap());
+                    }
+                }
+            }
+            None
+        })
+        .next();
```
Comment (mxinden) [inline swarm-derive/src/lib.rs:65]: This is mostly the same as the logic for `out_event`, right? What do you think of moving it into a function called in both places?

**Review by mxinden — APPROVED** [review]
One suggestion. Otherwise looks good to me.

File: swarm-derive/tests/test.rs [hunk 3]
```diff
@@ -133,6 +133,23 @@ fn custom_event() {
...
+    struct Foo {
+        ping: ping::Behaviour,
+        identify: identify::Behaviour,
+    }
+
+    #[allow(dead_code)]
+    fn foo() {
+        require_net_behaviour::<Foo>();
+    }
+}
```
Comment (elenaf9) [inline swarm-derive/tests/test.rs:151]: This test also already passes with current master (if you remove line 141). 
Correct me if I am wrong, but I believe the current `::libp2p` prefix imports libp2p from the crate level, which means that aliases in a module should not be an issue? @mxinden?

I think it is only a problem if the crate is imported in the `Cargo.toml` with a different name.
`serde` on the other hand states for their `crate` attribute that: 
> This is normally only applicable when invoking re-exported Serde derives from a public macro in a different crate.

I am not against adding this parameter, but could we add a test that fails on current master and would be fixed with this? 🙂

Comment (thomaseizinger) [reply swarm-derive/tests/test.rs:151]: As far as I understand, the issue is not clashing symbols but renaming the crate in your manifest.

To test that, we'd have to make a separate test crate with a renamed import.

This make me wonder, should we perhaps just require certain symbols to be present for the macro to work?

That would make https://github.com/libp2p/rust-libp2p/pull/3023 easier to implement.

Or perhaps we could make a seperate "derive" prelude? Like `libp2p::swarm_derive_prelude` that includes all the required symbols. Then we can add a configuration knob here that allows you to change where that prelude is coming from.

Comment (thomaseizinger) [reply swarm-derive/tests/test.rs:151]: > This make me wonder, should we perhaps just require certain symbols to be present for the macro to work?
> 
> That would make #3023 easier to implement.
> 
> Or perhaps we could make a seperate "derive" prelude? Like `libp2p::swarm_derive_prelude` that includes all the required symbols. Then we can add a configuration knob here that allows you to change where that prelude is coming from.

Actually, don't worry about it, I already have a solution to that.

**Comment by thomaseizinger** [comment]
If we end up shipping https://github.com/libp2p/rust-libp2p/pull/3055, then this PR is obsolete as the `prelude` option allows you to achieve the same thing via:

```rust
#[behaviour(prelude = "mylibp2p::swarm::derive_prelude")
```

**Comment by thomaseizinger** [comment]
Closing in favor of https://github.com/libp2p/rust-libp2p/pull/3055 which is about to be merged.

**Pull request closed without merging**